neuralmonkey/decoders/decoder.py

```diff
@@ -319,124 +374,202 @@ def get_attention_object(self, encoder, train_mode: bool):
 
         return self._runtime_attention_objects.get(encoder)
 
-    def step(self,
-             att_objects: List[BaseAttention],
-             input_: tf.Tensor,
-             prev_state: tf.Tensor,
-             prev_attns: List[tf.Tensor]):
+    def embed_input_symbol(self, *args) -> tf.Tensor:
+        loop_state = LoopState(*args)
 
-        with tf.variable_scope(self.step_scope):
-            cell = self._get_rnn_cell()
+        embedded_input = tf.nn.embedding_lookup(
+            self.embedding_matrix, loop_state.input_symbol)
 
-            # Merge input and previous attentions into one vector of the
-            # right size.
-            if self._attention_on_input:
-                x = linear([input_] + prev_attns, self.embedding_size)
-            else:
-                x = input_
-
-            # Run the RNN.
-            cell_output, state = cell(x, prev_state)
-
-            # Run the attention mechanism.
-            if self._rnn_cell_str == 'GRU':
-                attns = [a.attention(cell_output, prev_state, x)
-                         for a in att_objects]
-            elif self._rnn_cell_str == 'LSTM':
-                attns = [a.attention(cell_output, prev_state.c, x)
-                         for a in att_objects]
-            else:
-                raise ValueError("Unknown RNN cell.")
-
-            if self._conditional_gru and self._rnn_cell_str == "GRU":
-                cell_cond = self._get_conditional_gru_cell()
-                cond_input = tf.concat(attns, -1)
-                cell_output, state = cell_cond(cond_input, state,
-                                               scope="cond_gru_2_cell")
-
-            with tf.name_scope("rnn_output_projection"):
-                if attns:
-                    output = linear([cell_output] + attns,
-                                    cell.output_size,
-                                    scope="AttnOutputProjection")
+        return dropout(embedded_input, self.dropout_keep_prob, self.train_mode)
+
+    def input_plus_attention(self, *args) -> tf.Tensor:
+        """Merge input and previous attentions into one vector of the
+         right size.
+        """
+        loop_state = LoopState(*args)
+
+        embedded_input = self.embed_input_symbol(*loop_state)
+
+        return linear([embedded_input] + loop_state.prev_contexts,
+                      self.embedding_size)
+
+    def get_body(self,
+                 att_objects: List[BaseAttention],
+                 train_mode: bool,
+                 sample: bool = False) -> Callable[[LoopState], LoopState]:
```

Comment:
to bych nedával do tohohle pull requestu - stačí, že ještě neni správně ten beam search. Udělej kdyžtak issue na obecný používání higher-order funkcí. Minimálně ještě pro output_projection dekodéru, kde je sice udělanej arg pro decoder, ale nepoužívá se.